google/cloud/opentelemetry/trace_exporter.cc

```diff
@@ -0,0 +1,84 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/opentelemetry/trace_exporter.h"
+#include "google/cloud/trace/v2/trace_client.h"
+
+namespace google {
+namespace cloud {
+namespace otel {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+class TraceExporter final : public opentelemetry::sdk::trace::SpanExporter {
+ public:
+  explicit TraceExporter(Project project,
+                         std::shared_ptr<trace_v2::TraceServiceConnection> conn)
+      : project_(std::move(project)), client_(std::move(conn)) {}
+
+  std::unique_ptr<opentelemetry::sdk::trace::Recordable>
+  MakeRecordable() noexcept override {
+    return std::make_unique<otel_internal::Recordable>(project_);
+  }
+
+  opentelemetry::sdk::common::ExportResult Export(
+      opentelemetry::nostd::span<
+          std::unique_ptr<opentelemetry::sdk::trace::Recordable>> const&
+          spans) noexcept override {
+    google::devtools::cloudtrace::v2::BatchWriteSpansRequest request;
+    request.set_name(project_.FullName());
+    for (auto& recordable : spans) {
+      auto span = std::unique_ptr<otel_internal::Recordable>(
+          static_cast<otel_internal::Recordable*>(recordable.release()));
```

Comment:
> The `static_cast` could use another set of eyes.

I assume you can assert that every `opentelemetry::sdk::trace::Recordable` you see actually IS-A `otel_internal::Recordable`, in which case it looks fine.

The `Export()` interface is a little strange in that (I assume again) it is supposed to consume all the `unique_ptr`s, and yet it doesn't take them by value.

Reply:
Thanks. I was wondering if it was necessary to put the raw pointer back into a `std::unique_ptr`, when I dereference it immediately after.

I gave up though.

Reply:
Yes ... it needs to be put back into a `unique_ptr` (or deleted in another way) otherwise the moved-from span would be leaked.